# Multilingual Sentiment Annotation & QA

## Objective

This notebook performs the main annotation and quality assurance (QA)
process for the multilingual sentiment dataset.

The annotation follows the criteria defined in the annotation guideline
developed during the pilot annotation stage.

### Annotation Labels

- **Polarity:** Positive / Negative
- **Intensity:** Low / Medium / High
- **Languages:** English (EN), Chinese (ZH), Korean (KR)

### Workflow

1. Load the source dataset
2. Select an annotation batch
3. Perform manual annotation
4. Save annotation results
5. Conduct QA review
6. Analyze disagreements and edge cases
7. Update the annotation guideline when necessary

In [1]:
# AQ-02. Import libraries

import pandas as pd
from pathlib import Path

In [2]:
# AQ-03. Set project paths

BASE_DIR = Path.cwd().parent

DATA_DIR = BASE_DIR / "data"
RAW_DIR = DATA_DIR / "raw"
PROCESSED_DIR = DATA_DIR / "processed"

print("Base directory:", BASE_DIR)
print("Raw data directory:", RAW_DIR)
print("Processed data directory:", PROCESSED_DIR)

Base directory: c:\Users\user\OneDrive\Documents\data-labeling-portfolio\01_multilingual_sentiment
Raw data directory: c:\Users\user\OneDrive\Documents\data-labeling-portfolio\01_multilingual_sentiment\data\raw
Processed data directory: c:\Users\user\OneDrive\Documents\data-labeling-portfolio\01_multilingual_sentiment\data\processed


In [3]:
# AQ-04. Check available raw data files

for file in RAW_DIR.iterdir():
    print(file.name)

amazon_reviews_en_ground_truth_300.csv
amazon_reviews_zh_ground_truth_300.csv
nsmc_ko_ground_truth_300.csv
ratings_train.txt


In [4]:
# AQ-05. Load English source dataset

EN_FILE = RAW_DIR / "amazon_reviews_en_ground_truth_300.csv"

df_en = pd.read_csv(EN_FILE)

print("Dataset shape:", df_en.shape)
print("\nColumns:")
print(df_en.columns.tolist())

df_en.head()

Dataset shape: (300, 8)

Columns:
['review_id', 'product_id', 'reviewer_id', 'stars', 'review_body', 'review_title', 'language', 'product_category']


,review_id,product_id,reviewer_id,stars,review_body,review_title,language,product_category
0,en_0853851,product_en_0691102,reviewer_en_0203389,1,I did receive the item and it says it did arri...,I don’t recomend,en,other
1,en_0737292,product_en_0509265,reviewer_en_0357543,1,The difficulty I had with using these is that ...,Will block screen if there are screens on the ...,en,luggage
2,en_0574934,product_en_0898135,reviewer_en_0460742,1,Did not include the safety glove promised in t...,"Missing glove, dirty guard, customer service MIA",en,kitchen
3,en_0561880,product_en_0665297,reviewer_en_0481291,1,I received an email from Amazon.com stating th...,Received questionable email from Amazon.com pe...,en,other
4,en_0007287,product_en_0533382,reviewer_en_0542162,1,"They float. The bag says they don’t, but they ...",Floaters,en,lawn_and_garden


## AQ-06. English Annotation Batch 01

The first main annotation batch contains English reviews #30–49.

To reduce annotation bias, existing star ratings and identifying
metadata are excluded from the annotation view.

The annotator assigns:

- **Polarity:** Positive / Negative
- **Intensity:** Low / Medium / High

In [5]:
# AQ-06. Create English annotation batch #30-49

en_batch_01 = df_en.loc[
    30:49,
    ["review_id", "review_title", "review_body"]
].copy()

en_batch_01["polarity"] = ""
en_batch_01["intensity"] = ""

print("Batch size:", len(en_batch_01))

en_batch_01

Batch size: 20


,review_id,review_title,review_body,polarity,intensity
30,en_0328488,Unhappy with purchase,Seal was broken. Product looks different than ...,,
31,en_0462975,Don’t purchase,Product didn’t work as stated. Followed direct...,,
32,en_0565881,Spilled in shipping,The seeds were received quickly but were spill...,,
33,en_0636834,Terrible design,This design causes the fan to constantly move ...,,
34,en_0106633,One Star,Bought these and have had them up two weeks an...,,
35,en_0087364,Wireless charging stops after a few minutes of...,I bought this for the wireless charging capabi...,,
36,en_0867652,Unreliable,I never received the package!,,
37,en_0779152,One Star,I ordered a 2 pack and only received one pack!,,
38,en_0879650,Cheap Clasp-Lost Fitbit,I loved the look of this band but unfortunatel...,,
39,en_0635600,Not a supported accessory,"Works for a minute or so, then gets the error,...",,


## AQ-07. Manual Annotation — English Batch 01

Each review is evaluated independently based on its title and full review text.

Existing star ratings are intentionally hidden to minimize annotation bias.

In [6]:
# AQ-07. Display full review text for manual annotation

for idx, row in en_batch_01.iterrows():
    print(f"[{idx}]")
    print(f"Title: {row['review_title']}")
    print(f"Body : {row['review_body']}")
    print("-" * 80)

[30]
Title: Unhappy with purchase
Body : Seal was broken. Product looks different than prior purchase.
--------------------------------------------------------------------------------
[31]
Title: Don’t purchase
Body : Product didn’t work as stated. Followed directions and it’s currently the 8th day and haven’t seen any peeling or difference. Will not buy again. Very disappointed. Will be buying Baby Foot.
--------------------------------------------------------------------------------
[32]
Title: Spilled in shipping
Body : The seeds were received quickly but were spilled out into shipping envelope.
--------------------------------------------------------------------------------
[33]
Title: Terrible design
Body : This design causes the fan to constantly move away from you and swirl around. Buy one that is enclosed all the way around.
--------------------------------------------------------------------------------
[34]
Title: One Star
Body : Bought these and have had them up two weeks an

## AQ-08. QA Review — English Batch 01

The manual annotations were independently reviewed using Annotation Guideline v0.2.

The QA process evaluates:

- **Polarity agreement**
- **Intensity agreement**
- **Disagreement reason**
- Potential edge cases for future guideline refinement

> **Key QA principle:**  
> The severity of the reported problem does not necessarily equal the
> intensity of the sentiment expressed in the review.

In [7]:
# AQ-08. Record QA results

qa_results_en_01 = pd.DataFrame({
    "review_id": en_batch_01["review_id"].values,

    "annotator_polarity": ["Negative"] * 20,
    "annotator_intensity": [
        "High", "High", "High", "High", "High",
        "High", "High", "High", "High", "High",
        "High", "High", "High", "High", "High",
        "High", "High", "High", "High", "High"
    ],

    "qa_polarity": ["Negative"] * 20,
    "qa_intensity": [
        "Medium", "High", "Low", "High", "Medium",
        "High", "Medium", "Medium", "High", "High",
        "Medium", "High", "Medium", "Medium", "High",
        "Medium", "Medium", "High", "Medium", "High"
    ]
})

qa_results_en_01["polarity_agreement"] = (
    qa_results_en_01["annotator_polarity"]
    == qa_results_en_01["qa_polarity"]
)

qa_results_en_01["intensity_agreement"] = (
    qa_results_en_01["annotator_intensity"]
    == qa_results_en_01["qa_intensity"]
)

qa_results_en_01

,review_id,annotator_polarity,annotator_intensity,qa_polarity,qa_intensity,polarity_agreement,intensity_agreement
0,en_0328488,Negative,High,Negative,Medium,True,False
1,en_0462975,Negative,High,Negative,High,True,True
2,en_0565881,Negative,High,Negative,Low,True,False
3,en_0636834,Negative,High,Negative,High,True,True
4,en_0106633,Negative,High,Negative,Medium,True,False
5,en_0087364,Negative,High,Negative,High,True,True
6,en_0867652,Negative,High,Negative,Medium,True,False
7,en_0779152,Negative,High,Negative,Medium,True,False
8,en_0879650,Negative,High,Negative,High,True,True
9,en_0635600,Negative,High,Negative,High,True,True


## AQ-09. Annotation Agreement Metrics

Agreement rates are calculated separately for polarity and sentiment intensity.

This helps identify whether annotation inconsistencies originate from
sentiment direction or from differences in perceived emotional strength.

In [8]:
# AQ-09. Calculate agreement metrics

polarity_agreement_rate = (
    qa_results_en_01["polarity_agreement"].mean() * 100
)

intensity_agreement_rate = (
    qa_results_en_01["intensity_agreement"].mean() * 100
)

overall_exact_agreement = (
    (
        qa_results_en_01["polarity_agreement"]
        & qa_results_en_01["intensity_agreement"]
    ).mean() * 100
)

print(f"Polarity agreement : {polarity_agreement_rate:.1f}%")
print(f"Intensity agreement: {intensity_agreement_rate:.1f}%")
print(f"Exact label agreement: {overall_exact_agreement:.1f}%")

Polarity agreement : 100.0%
Intensity agreement: 45.0%
Exact label agreement: 45.0%


In [9]:
# Extract intensity disagreements

intensity_disagreements = qa_results_en_01[
    ~qa_results_en_01["intensity_agreement"]
].copy()

print("Number of intensity disagreements:", len(intensity_disagreements))

intensity_disagreements

Number of intensity disagreements: 11


,review_id,annotator_polarity,annotator_intensity,qa_polarity,qa_intensity,polarity_agreement,intensity_agreement
0,en_0328488,Negative,High,Negative,Medium,True,False
2,en_0565881,Negative,High,Negative,Low,True,False
4,en_0106633,Negative,High,Negative,Medium,True,False
6,en_0867652,Negative,High,Negative,Medium,True,False
7,en_0779152,Negative,High,Negative,Medium,True,False
10,en_0863454,Negative,High,Negative,Medium,True,False
12,en_0378808,Negative,High,Negative,Medium,True,False
13,en_0167781,Negative,High,Negative,Medium,True,False
15,en_0066766,Negative,High,Negative,Medium,True,False
16,en_0229378,Negative,High,Negative,Medium,True,False


## AQ-10. Disagreement Analysis — English Batch 01

### Key Finding

Polarity agreement reached **100%**, indicating that the annotator
consistently identified the direction of sentiment.

However, intensity agreement was **45%**, with 11 disagreements.
All intensity disagreements resulted from the annotator assigning
**High** intensity where the QA review assigned **Medium or Low**.

### Main Disagreement Pattern

The primary source of disagreement was the distinction between:

- **Severity of the reported problem**
- **Intensity of the sentiment expressed in the text**

A serious product or delivery problem does not automatically indicate
high sentiment intensity.

### Examples

- **#32 — Negative / Low**  
  "The seeds were received quickly but were spilled out into shipping envelope."

  The review reports a negative event but contains little explicit
  emotional language.

- **#42 — Negative / Medium**  
  "I did not receive order."

  Non-delivery is a serious issue, but the reviewer communicates it
  briefly without strongly intensified emotional language.

- **#31 — Negative / High**  
  "Will not buy again. Very disappointed."

  Explicit disappointment and rejection of future purchase provide
  clear evidence of high negative intensity.

### QA Insight

Intensity should primarily reflect the **strength of sentiment expressed
through language**, rather than the objective severity of the underlying
product or service problem.

## AQ-11. Final Label Adjudication — English Batch 01

After QA review and disagreement analysis, final labels are assigned
through adjudication.

The original annotator labels and QA labels are preserved separately
to maintain a transparent annotation history.

For this batch, the final labels follow the QA-reviewed decisions
after resolving the identified intensity disagreements.

In [10]:
# AQ-11. Assign final labels after QA adjudication

qa_results_en_01["final_polarity"] = qa_results_en_01["qa_polarity"]
qa_results_en_01["final_intensity"] = qa_results_en_01["qa_intensity"]

qa_results_en_01[
    [
        "review_id",
        "annotator_polarity",
        "annotator_intensity",
        "qa_polarity",
        "qa_intensity",
        "final_polarity",
        "final_intensity"
    ]
]

,review_id,annotator_polarity,annotator_intensity,qa_polarity,qa_intensity,final_polarity,final_intensity
0,en_0328488,Negative,High,Negative,Medium,Negative,Medium
1,en_0462975,Negative,High,Negative,High,Negative,High
2,en_0565881,Negative,High,Negative,Low,Negative,Low
3,en_0636834,Negative,High,Negative,High,Negative,High
4,en_0106633,Negative,High,Negative,Medium,Negative,Medium
5,en_0087364,Negative,High,Negative,High,Negative,High
6,en_0867652,Negative,High,Negative,Medium,Negative,Medium
7,en_0779152,Negative,High,Negative,Medium,Negative,Medium
8,en_0879650,Negative,High,Negative,High,Negative,High
9,en_0635600,Negative,High,Negative,High,Negative,High


In [11]:
# AQ-12. Create finalized annotation dataset

final_en_batch_01 = en_batch_01[
    ["review_id", "review_title", "review_body"]
].merge(
    qa_results_en_01[
        ["review_id", "final_polarity", "final_intensity"]
    ],
    on="review_id",
    how="left"
)

final_en_batch_01["language"] = "en"

final_en_batch_01

,review_id,review_title,review_body,final_polarity,final_intensity,language
0,en_0328488,Unhappy with purchase,Seal was broken. Product looks different than ...,Negative,Medium,en
1,en_0462975,Don’t purchase,Product didn’t work as stated. Followed direct...,Negative,High,en
2,en_0565881,Spilled in shipping,The seeds were received quickly but were spill...,Negative,Low,en
3,en_0636834,Terrible design,This design causes the fan to constantly move ...,Negative,High,en
4,en_0106633,One Star,Bought these and have had them up two weeks an...,Negative,Medium,en
5,en_0087364,Wireless charging stops after a few minutes of...,I bought this for the wireless charging capabi...,Negative,High,en
6,en_0867652,Unreliable,I never received the package!,Negative,Medium,en
7,en_0779152,One Star,I ordered a 2 pack and only received one pack!,Negative,Medium,en
8,en_0879650,Cheap Clasp-Lost Fitbit,I loved the look of this band but unfortunatel...,Negative,High,en
9,en_0635600,Not a supported accessory,"Works for a minute or so, then gets the error,...",Negative,High,en


## AQ-13. Save English Batch 01 Results

Two separate outputs are saved:

1. **Final annotation dataset** — contains the adjudicated labels used as the final dataset.
2. **QA audit dataset** — preserves annotator labels, QA labels, agreement results,
   and final labels for quality tracking and reproducibility.

In [12]:
# AQ-13. Save finalized annotation and QA audit files

ANNOTATION_FILE = PROCESSED_DIR / "annotation_en_batch_01.csv"
QA_FILE = PROCESSED_DIR / "qa_en_batch_01.csv"

final_en_batch_01.to_csv(
    ANNOTATION_FILE,
    index=False,
    encoding="utf-8-sig"
)

qa_results_en_01.to_csv(
    QA_FILE,
    index=False,
    encoding="utf-8-sig"
)

print("Saved annotation file:", ANNOTATION_FILE)
print("Saved QA file:", QA_FILE)

Saved annotation file: c:\Users\user\OneDrive\Documents\data-labeling-portfolio\01_multilingual_sentiment\data\processed\annotation_en_batch_01.csv
Saved QA file: c:\Users\user\OneDrive\Documents\data-labeling-portfolio\01_multilingual_sentiment\data\processed\qa_en_batch_01.csv


In [13]:
# AQ-14. Verify saved files

saved_annotation = pd.read_csv(ANNOTATION_FILE)
saved_qa = pd.read_csv(QA_FILE)

print("Annotation shape:", saved_annotation.shape)
print("QA shape:", saved_qa.shape)

print("\nMissing final labels:")
print(
    saved_annotation[
        ["final_polarity", "final_intensity"]
    ].isna().sum()
)

print("\nAnnotation file preview:")
display(saved_annotation.head())

Annotation shape: (20, 6)
QA shape: (20, 9)

Missing final labels:
final_polarity     0
final_intensity    0
dtype: int64

Annotation file preview:


,review_id,review_title,review_body,final_polarity,final_intensity,language
0,en_0328488,Unhappy with purchase,Seal was broken. Product looks different than ...,Negative,Medium,en
1,en_0462975,Don’t purchase,Product didn’t work as stated. Followed direct...,Negative,High,en
2,en_0565881,Spilled in shipping,The seeds were received quickly but were spill...,Negative,Low,en
3,en_0636834,Terrible design,This design causes the fan to constantly move ...,Negative,High,en
4,en_0106633,One Star,Bought these and have had them up two weeks an...,Negative,Medium,en


## AQ-15. English Annotation Batch 02

English reviews #50–69 are annotated using **Annotation Guideline v0.3**.

This batch is used to evaluate whether the refined intensity criteria
improve annotation consistency following the QA findings from Batch 01.

### Focus

- Separate **problem severity** from **expressed sentiment intensity**
- Use linguistic evidence when assigning High intensity
- Apply Low / Medium / High criteria consistently

In [14]:
# AQ-15. Create English annotation batch #50-69

en_batch_02 = df_en.loc[
    50:69,
    ["review_id", "review_title", "review_body"]
].copy()

en_batch_02["polarity"] = ""
en_batch_02["intensity"] = ""

print("Batch size:", len(en_batch_02))

en_batch_02

Batch size: 20


,review_id,review_title,review_body,polarity,intensity
50,en_0452426,BAD,arrived in very bad condition while the packag...,,
51,en_0591222,Does not seem durable - snap top already cracked.,Open the box and discover a top side snap alre...,,
52,en_0243936,"DID NOT WORK, COULD NOT BE CODED","PURCHASE WITH UNDERSTANDING THAT THE KEY WORK,...",,
53,en_0712253,Shipped wrong product,Order was for UNSCENTED. Received scented!,,
54,en_0186916,Hard to believe!,UPDATE: The seller send out another tent for m...,,
55,en_0236997,Did not work,I bought this to sharpen my Wustof Ikon knife ...,,
56,en_0106298,One Star,I ordered xxx- large and they sent size medium.,,
57,en_0210313,What's wrong with this app,2hrs trying to sign.....kept telling me wrong ...,,
58,en_0527053,Quality was not good.,Feels cheap. It fits though.,,
59,en_0701693,One Star,"Just the sticker , scam",,


In [15]:
# AQ-16. Display full review text for manual annotation

for idx, row in en_batch_02.iterrows():
    print(f"[{idx}]")
    print(f"Title: {row['review_title']}")
    print(f"Body : {row['review_body']}")
    print("-" * 80)

[50]
Title: BAD
Body : arrived in very bad condition while the packaging was impeccable
--------------------------------------------------------------------------------
[51]
Title: Does not seem durable - snap top already cracked.
Body : Open the box and discover a top side snap already cracked. I want to give this a chance but wonder about the durability of this product now.
--------------------------------------------------------------------------------
[52]
Title: DID NOT WORK, COULD NOT BE CODED
Body : PURCHASE WITH UNDERSTANDING THAT THE KEY WORK, FOR MY CAR, YEAR AND MAKE..... BROUGHT TO HAVE IT CUT AND KEYED.... CUT BUT COULD NOT BE CODED... SO USELESS..... BEST GO TO DEALER... IN LONG RUN.... I HAD LOST MONEY TRYING TO GET IT TO WORK... SO LOST MONEY... AND STILL DO NOT HAVE THE AUTO KEY I NEED......
--------------------------------------------------------------------------------
[53]
Title: Shipped wrong product
Body : Order was for UNSCENTED. Received scented!
---------------

In [16]:
# AQ-17. Record QA results — English Batch 02

qa_results_en_02 = pd.DataFrame({
    "review_id": en_batch_02["review_id"].values,

    "annotator_polarity": ["Negative"] * 20,

    "annotator_intensity": [
        "Medium", "Medium", "High", "Medium", "High",
        "High", "Medium", "High", "Medium", "High",
        "Medium", "Medium", "Medium", "High", "Medium",
        "High", "Medium", "Medium", "High", "Medium"
    ],

    "qa_polarity": ["Negative"] * 20,

    "qa_intensity": [
        "Medium", "Medium", "High", "Medium", "High",
        "High", "Medium", "High", "Medium", "High",
        "Medium", "Medium", "Medium", "Medium", "Medium",
        "Medium", "Medium", "Medium", "High", "Medium"
    ]
})

qa_results_en_02["polarity_agreement"] = (
    qa_results_en_02["annotator_polarity"]
    == qa_results_en_02["qa_polarity"]
)

qa_results_en_02["intensity_agreement"] = (
    qa_results_en_02["annotator_intensity"]
    == qa_results_en_02["qa_intensity"]
)

qa_results_en_02

,review_id,annotator_polarity,annotator_intensity,qa_polarity,qa_intensity,polarity_agreement,intensity_agreement
0,en_0452426,Negative,Medium,Negative,Medium,True,True
1,en_0591222,Negative,Medium,Negative,Medium,True,True
2,en_0243936,Negative,High,Negative,High,True,True
3,en_0712253,Negative,Medium,Negative,Medium,True,True
4,en_0186916,Negative,High,Negative,High,True,True
5,en_0236997,Negative,High,Negative,High,True,True
6,en_0106298,Negative,Medium,Negative,Medium,True,True
7,en_0210313,Negative,High,Negative,High,True,True
8,en_0527053,Negative,Medium,Negative,Medium,True,True
9,en_0701693,Negative,High,Negative,High,True,True


## AQ-18. QA Improvement Comparison

Annotation agreement is compared between English Batch 01 and Batch 02
to evaluate the effect of the guideline refinement.

Batch 01 was annotated using **Guideline v0.2**, while Batch 02 was
annotated after the intensity criteria were refined in **Guideline v0.3**.

In [17]:
# AQ-18. Compare QA metrics between English batches

comparison = pd.DataFrame({
    "Batch": ["EN Batch 01", "EN Batch 02"],
    "Guideline": ["v0.2", "v0.3"],
    "Polarity Agreement (%)": [
        qa_results_en_01["polarity_agreement"].mean() * 100,
        qa_results_en_02["polarity_agreement"].mean() * 100
    ],
    "Intensity Agreement (%)": [
        qa_results_en_01["intensity_agreement"].mean() * 100,
        qa_results_en_02["intensity_agreement"].mean() * 100
    ],
    "Exact Agreement (%)": [
        (
            qa_results_en_01["polarity_agreement"]
            & qa_results_en_01["intensity_agreement"]
        ).mean() * 100,
        (
            qa_results_en_02["polarity_agreement"]
            & qa_results_en_02["intensity_agreement"]
        ).mean() * 100
    ]
})

comparison.round(1)

,Batch,Guideline,Polarity Agreement (%),Intensity Agreement (%),Exact Agreement (%)
0,EN Batch 01,v0.2,100.0,45.0,45.0
1,EN Batch 02,v0.3,100.0,90.0,90.0


### Interpretation

- Polarity agreement remained stable at **100%** across both batches.
- Intensity agreement increased from **45% to 90%** after the guideline refinement.
- Exact label agreement consequently increased from **45% to 90%**.
- The largest improvement came from distinguishing **problem severity**
  from the **strength of sentiment expressed in the text**.
- Two remaining disagreements (#63 and #65) involved the boundary between
  Medium and High intensity.

> Because each batch contains only 20 reviews, this improvement should be
> interpreted as preliminary evidence of increased annotation consistency,
> rather than a statistically conclusive result.

## AQ-19. Final Label Adjudication — English Batch 02

Following QA review, disagreements were adjudicated while preserving
the original annotator and QA labels.

Two intensity disagreements (#63 and #65) were resolved as **Medium**
based on Guideline v0.3.

In [18]:
# AQ-19. Assign final labels after QA adjudication

qa_results_en_02["final_polarity"] = qa_results_en_02["qa_polarity"]
qa_results_en_02["final_intensity"] = qa_results_en_02["qa_intensity"]

qa_results_en_02[
    [
        "review_id",
        "annotator_polarity",
        "annotator_intensity",
        "qa_polarity",
        "qa_intensity",
        "final_polarity",
        "final_intensity"
    ]
]

,review_id,annotator_polarity,annotator_intensity,qa_polarity,qa_intensity,final_polarity,final_intensity
0,en_0452426,Negative,Medium,Negative,Medium,Negative,Medium
1,en_0591222,Negative,Medium,Negative,Medium,Negative,Medium
2,en_0243936,Negative,High,Negative,High,Negative,High
3,en_0712253,Negative,Medium,Negative,Medium,Negative,Medium
4,en_0186916,Negative,High,Negative,High,Negative,High
5,en_0236997,Negative,High,Negative,High,Negative,High
6,en_0106298,Negative,Medium,Negative,Medium,Negative,Medium
7,en_0210313,Negative,High,Negative,High,Negative,High
8,en_0527053,Negative,Medium,Negative,Medium,Negative,Medium
9,en_0701693,Negative,High,Negative,High,Negative,High


In [19]:
# AQ-20. Create finalized annotation dataset — Batch 02

final_en_batch_02 = en_batch_02[
    ["review_id", "review_title", "review_body"]
].merge(
    qa_results_en_02[
        ["review_id", "final_polarity", "final_intensity"]
    ],
    on="review_id",
    how="left"
)

final_en_batch_02["language"] = "en"

final_en_batch_02

,review_id,review_title,review_body,final_polarity,final_intensity,language
0,en_0452426,BAD,arrived in very bad condition while the packag...,Negative,Medium,en
1,en_0591222,Does not seem durable - snap top already cracked.,Open the box and discover a top side snap alre...,Negative,Medium,en
2,en_0243936,"DID NOT WORK, COULD NOT BE CODED","PURCHASE WITH UNDERSTANDING THAT THE KEY WORK,...",Negative,High,en
3,en_0712253,Shipped wrong product,Order was for UNSCENTED. Received scented!,Negative,Medium,en
4,en_0186916,Hard to believe!,UPDATE: The seller send out another tent for m...,Negative,High,en
5,en_0236997,Did not work,I bought this to sharpen my Wustof Ikon knife ...,Negative,High,en
6,en_0106298,One Star,I ordered xxx- large and they sent size medium.,Negative,Medium,en
7,en_0210313,What's wrong with this app,2hrs trying to sign.....kept telling me wrong ...,Negative,High,en
8,en_0527053,Quality was not good.,Feels cheap. It fits though.,Negative,Medium,en
9,en_0701693,One Star,"Just the sticker , scam",Negative,High,en


In [20]:
# AQ-21. Save English Batch 02 results

ANNOTATION_FILE_02 = PROCESSED_DIR / "annotation_en_batch_02.csv"
QA_FILE_02 = PROCESSED_DIR / "qa_en_batch_02.csv"

final_en_batch_02.to_csv(
    ANNOTATION_FILE_02,
    index=False,
    encoding="utf-8-sig"
)

qa_results_en_02.to_csv(
    QA_FILE_02,
    index=False,
    encoding="utf-8-sig"
)

print("Saved annotation file:", ANNOTATION_FILE_02)
print("Saved QA file:", QA_FILE_02)

Saved annotation file: c:\Users\user\OneDrive\Documents\data-labeling-portfolio\01_multilingual_sentiment\data\processed\annotation_en_batch_02.csv
Saved QA file: c:\Users\user\OneDrive\Documents\data-labeling-portfolio\01_multilingual_sentiment\data\processed\qa_en_batch_02.csv


In [21]:
# AQ-22. Verify saved Batch 02 files

saved_annotation_02 = pd.read_csv(ANNOTATION_FILE_02)
saved_qa_02 = pd.read_csv(QA_FILE_02)

print("Annotation shape:", saved_annotation_02.shape)
print("QA shape:", saved_qa_02.shape)

print("\nMissing final labels:")
print(
    saved_annotation_02[
        ["final_polarity", "final_intensity"]
    ].isna().sum()
)

print("\nFinal intensity distribution:")
print(
    saved_annotation_02["final_intensity"]
    .value_counts()
)

display(saved_annotation_02.head())

Annotation shape: (20, 6)
QA shape: (20, 9)

Missing final labels:
final_polarity     0
final_intensity    0
dtype: int64

Final intensity distribution:
final_intensity
Medium    14
High       6
Name: count, dtype: int64


,review_id,review_title,review_body,final_polarity,final_intensity,language
0,en_0452426,BAD,arrived in very bad condition while the packag...,Negative,Medium,en
1,en_0591222,Does not seem durable - snap top already cracked.,Open the box and discover a top side snap alre...,Negative,Medium,en
2,en_0243936,"DID NOT WORK, COULD NOT BE CODED","PURCHASE WITH UNDERSTANDING THAT THE KEY WORK,...",Negative,High,en
3,en_0712253,Shipped wrong product,Order was for UNSCENTED. Received scented!,Negative,Medium,en
4,en_0186916,Hard to believe!,UPDATE: The seller send out another tent for m...,Negative,High,en


## AQ-23. English Annotation Batch 03

English reviews #70–89 are annotated using **Annotation Guideline v0.3**.

This batch continues the evaluation of annotation consistency after
the intensity criteria refinement introduced following Batch 01.

### Annotation Focus

- Evaluate sentiment direction independently.
- Distinguish problem severity from expressed sentiment intensity.
- Assign High intensity only when supported by clear linguistic evidence.
- Apply the same criteria consistently across reviews.

In [22]:
# AQ-23. Create English annotation batch #70-89

en_batch_03 = df_en.loc[
    70:89,
    ["review_id", "review_title", "review_body"]
].copy()

en_batch_03["polarity"] = ""
en_batch_03["intensity"] = ""

print("Batch size:", len(en_batch_03))

en_batch_03

Batch size: 20


,review_id,review_title,review_body,polarity,intensity
70,en_0941733,Sizing,Ordered the size I always buy and it’s too sma...,,
71,en_0021765,Does not work!,This shampoo did not remove orange from my hai...,,
72,en_0422019,Too Big,Absolutely Giant. Order in the next size down....,,
73,en_0886190,Not safe due to chemical smells.,"Extremely strong chemical smell, made my pet r...",,
74,en_0375426,Didn’t do anything,I ordered this bc my 9 year old son had a wart...,,
75,en_0577473,"Initially seems good, then disappointing.",Printing looks very nice paper very nice quali...,,
76,en_0900485,Extremely small coil hair ties.,These are tiny or mini(not mentioned in descri...,,
77,en_0188946,Not exactly what I expected,I think the idea has merit but lacked in perfo...,,
78,en_0713362,Cool item but very poor packaging,I was hoping to give this item as a gift to so...,,
79,en_0256854,Scratch material in the waist area,"Whatever material they use for ""tummy control""...",,


In [23]:
# AQ-24. Display full review text for manual annotation

for idx, row in en_batch_03.iterrows():
    print(f"[{idx}]")
    print(f"Title: {row['review_title']}")
    print(f"Body : {row['review_body']}")
    print("-" * 80)

[70]
Title: Sizing
Body : Ordered the size I always buy and it’s too small, disappointed in the sizing of the garment, other than that it a very nice garment.
--------------------------------------------------------------------------------
[71]
Title: Does not work!
Body : This shampoo did not remove orange from my hair, in fact, it brought back orange! It stripped the dark ash blonde dye from my hair and left it back in its orange state. My hair felt very soft in the morning, that’s why I’m giving it 2 stars. But it did not tone the orange out at all, it made my hair even more orange. I will use it a few more times and if it starts working I will update my review.
--------------------------------------------------------------------------------
[72]
Title: Too Big
Body : Absolutely Giant. Order in the next size down. Or dont order this at all...
--------------------------------------------------------------------------------
[73]
Title: Not safe due to chemical smells.
Body : Extremely

## AQ-25. Polarity Edge Case Analysis — English Batch 03

During Batch 03, several reviews contained both positive and negative
statements or were expressed in an ambiguous or restrained manner.

The annotator initially used **Mixed, Neutral, or Unclear** for six reviews.
However, the current annotation schema defines polarity as a binary label:

- Positive
- Negative

Therefore, these cases require adjudication based on the review's
**dominant evaluative direction**, rather than the presence of both
positive and negative statements.

### Edge Case 1 — Mixed Sentiment

A review may contain both positive and negative statements without
requiring a separate Mixed label.

The final polarity should reflect the **dominant evaluation or overall
conclusion of the review**.

**Example — #70**

> "Ordered the size I always buy and it’s too small, disappointed in the
> sizing ... other than that it a very nice garment."

Positive evidence:
- "very nice garment"

Negative evidence:
- incorrect sizing
- explicit disappointment

**Adjudication: Negative / Medium**

Although positive product qualities are acknowledged, the main purpose
and conclusion of the review concern dissatisfaction with sizing.

---

**Example — #78**

> "Cool item but very poor packaging"

Positive evidence:
- "really fun idea"
- "would have made a great gift"

Negative evidence:
- packaging arrived torn and dirty
- reviewer could no longer use it as the intended gift

**Adjudication: Negative / Medium**

The product concept is evaluated positively, but the actual purchase
experience failed its intended purpose.

### Edge Case 2 — Neutral vs. Negative

A review should not be labeled Neutral simply because it contains
positive observations alongside negative feedback.

When the reviewer reports clear dissatisfaction, product failure,
or recommends caution or an alternative product, the review has
an evaluative direction.

**Example — #71**

The reviewer notes that the shampoo made the hair feel soft, but repeatedly
states that the product failed its primary purpose and made the original
problem worse.

**Adjudication: Negative / High**

---

**Example — #86**

The reviewer acknowledges affordability and possible limited use cases,
but reports repeated disconnections, product failure, and recommends
considering another product.

**Adjudication: Negative / Medium**

### Edge Case 3 — Short Text vs. Unclear Sentiment

A short review is not automatically unclear.

If the available text contains sufficient evaluative information,
polarity should still be assigned.

**Example — #85**

> "See through- the thin material"

Although extremely short, the statement clearly criticizes the material
quality.

**Adjudication: Negative / Low**

Use Unclear only when the text genuinely lacks enough information to
determine sentiment direction.

## AQ-26. Polarity Decision Rule

When a review contains mixed, balanced, or limited information:

1. Identify all positive and negative evidence.
2. Determine the primary product or service issue being evaluated.
3. Examine the reviewer's overall conclusion or behavioral response.
4. Assign the polarity that best represents the **dominant evaluative direction**.

### Important Distinctions

- Mixed content ≠ Mixed label
- Balanced language ≠ Neutral sentiment
- Short text ≠ Unclear sentiment

Under the current binary annotation schema, Positive or Negative should
be assigned whenever the dominant sentiment can reasonably be determined.

## AQ-27. QA Review — English Batch 03

Batch 03 introduced new polarity edge cases involving mixed, neutral,
and unclear interpretations.

QA evaluates both:

- **Polarity agreement**
- **Intensity agreement**

The original annotator labels are preserved before adjudication so that
disagreement patterns can be analyzed independently.

In [24]:
# AQ-27. Record QA results — English Batch 03

qa_results_en_03 = pd.DataFrame({
    "review_id": en_batch_03["review_id"].values,

    "annotator_polarity": [
        "Mixed", "Neutral", "Negative", "Negative", "Negative",
        "Negative", "Negative", "Negative", "Mixed", "Negative",
        "Negative", "Negative", "Negative", "Negative", "Negative",
        "Unclear", "Neutral", "Negative", "Mixed", "Negative"
    ],

    "annotator_intensity": [
        "Medium", "Medium", "Low", "Medium", "Medium",
        "Medium", "Medium", "Medium", "Medium", "Medium",
        "High", "Medium", "Medium", "Medium", "Medium",
        "Low", "Medium", "High", "Medium", "Medium"
    ],

    "qa_polarity": ["Negative"] * 20,

    "qa_intensity": [
        "Medium", "High", "High", "High", "Medium",
        "Medium", "Medium", "Medium", "Medium", "Medium",
        "High", "Medium", "Medium", "High", "Medium",
        "Low", "Medium", "High", "Medium", "Medium"
    ]
})

qa_results_en_03["polarity_agreement"] = (
    qa_results_en_03["annotator_polarity"]
    == qa_results_en_03["qa_polarity"]
)

qa_results_en_03["intensity_agreement"] = (
    qa_results_en_03["annotator_intensity"]
    == qa_results_en_03["qa_intensity"]
)

qa_results_en_03

,review_id,annotator_polarity,annotator_intensity,qa_polarity,qa_intensity,polarity_agreement,intensity_agreement
0,en_0941733,Mixed,Medium,Negative,Medium,False,True
1,en_0021765,Neutral,Medium,Negative,High,False,False
2,en_0422019,Negative,Low,Negative,High,True,False
3,en_0886190,Negative,Medium,Negative,High,True,False
4,en_0375426,Negative,Medium,Negative,Medium,True,True
5,en_0577473,Negative,Medium,Negative,Medium,True,True
6,en_0900485,Negative,Medium,Negative,Medium,True,True
7,en_0188946,Negative,Medium,Negative,Medium,True,True
8,en_0713362,Mixed,Medium,Negative,Medium,False,True
9,en_0256854,Negative,Medium,Negative,Medium,True,True


In [25]:
# AQ-28. Calculate Batch 03 QA metrics

polarity_agreement_03 = (
    qa_results_en_03["polarity_agreement"].mean() * 100
)

intensity_agreement_03 = (
    qa_results_en_03["intensity_agreement"].mean() * 100
)

exact_agreement_03 = (
    (
        qa_results_en_03["polarity_agreement"]
        & qa_results_en_03["intensity_agreement"]
    ).mean() * 100
)

print(f"Polarity agreement : {polarity_agreement_03:.1f}%")
print(f"Intensity agreement: {intensity_agreement_03:.1f}%")
print(f"Exact label agreement: {exact_agreement_03:.1f}%")

Polarity agreement : 70.0%
Intensity agreement: 80.0%
Exact label agreement: 55.0%


In [26]:
# AQ-29. Extract Batch 03 disagreements

polarity_disagreements_03 = qa_results_en_03[
    ~qa_results_en_03["polarity_agreement"]
].copy()

intensity_disagreements_03 = qa_results_en_03[
    ~qa_results_en_03["intensity_agreement"]
].copy()

print(
    "Polarity disagreements:",
    len(polarity_disagreements_03)
)

print(
    "Intensity disagreements:",
    len(intensity_disagreements_03)
)

print("\nPolarity disagreement IDs:")
print(polarity_disagreements_03["review_id"].tolist())

print("\nIntensity disagreement IDs:")
print(intensity_disagreements_03["review_id"].tolist())

Polarity disagreements: 6
Intensity disagreements: 4

Polarity disagreement IDs:
['en_0941733', 'en_0021765', 'en_0713362', 'en_0461803', 'en_0597009', 'en_0502771']

Intensity disagreement IDs:
['en_0021765', 'en_0422019', 'en_0886190', 'en_0876945']


## AQ-30. Final Label Adjudication — English Batch 03

Following QA review and polarity edge-case analysis, all disagreements
were adjudicated using the current binary polarity schema.

Original annotator labels are preserved in the QA audit dataset,
while QA-reviewed decisions are used as the final labels.

In [27]:
# AQ-30. Assign final labels after QA adjudication

qa_results_en_03["final_polarity"] = qa_results_en_03["qa_polarity"]
qa_results_en_03["final_intensity"] = qa_results_en_03["qa_intensity"]

qa_results_en_03[
    [
        "review_id",
        "annotator_polarity",
        "annotator_intensity",
        "qa_polarity",
        "qa_intensity",
        "final_polarity",
        "final_intensity"
    ]
]

,review_id,annotator_polarity,annotator_intensity,qa_polarity,qa_intensity,final_polarity,final_intensity
0,en_0941733,Mixed,Medium,Negative,Medium,Negative,Medium
1,en_0021765,Neutral,Medium,Negative,High,Negative,High
2,en_0422019,Negative,Low,Negative,High,Negative,High
3,en_0886190,Negative,Medium,Negative,High,Negative,High
4,en_0375426,Negative,Medium,Negative,Medium,Negative,Medium
5,en_0577473,Negative,Medium,Negative,Medium,Negative,Medium
6,en_0900485,Negative,Medium,Negative,Medium,Negative,Medium
7,en_0188946,Negative,Medium,Negative,Medium,Negative,Medium
8,en_0713362,Mixed,Medium,Negative,Medium,Negative,Medium
9,en_0256854,Negative,Medium,Negative,Medium,Negative,Medium


In [28]:
# AQ-31. Create finalized annotation dataset — Batch 03

final_en_batch_03 = en_batch_03[
    ["review_id", "review_title", "review_body"]
].merge(
    qa_results_en_03[
        ["review_id", "final_polarity", "final_intensity"]
    ],
    on="review_id",
    how="left"
)

final_en_batch_03["language"] = "en"

final_en_batch_03

,review_id,review_title,review_body,final_polarity,final_intensity,language
0,en_0941733,Sizing,Ordered the size I always buy and it’s too sma...,Negative,Medium,en
1,en_0021765,Does not work!,This shampoo did not remove orange from my hai...,Negative,High,en
2,en_0422019,Too Big,Absolutely Giant. Order in the next size down....,Negative,High,en
3,en_0886190,Not safe due to chemical smells.,"Extremely strong chemical smell, made my pet r...",Negative,High,en
4,en_0375426,Didn’t do anything,I ordered this bc my 9 year old son had a wart...,Negative,Medium,en
5,en_0577473,"Initially seems good, then disappointing.",Printing looks very nice paper very nice quali...,Negative,Medium,en
6,en_0900485,Extremely small coil hair ties.,These are tiny or mini(not mentioned in descri...,Negative,Medium,en
7,en_0188946,Not exactly what I expected,I think the idea has merit but lacked in perfo...,Negative,Medium,en
8,en_0713362,Cool item but very poor packaging,I was hoping to give this item as a gift to so...,Negative,Medium,en
9,en_0256854,Scratch material in the waist area,"Whatever material they use for ""tummy control""...",Negative,Medium,en


In [29]:
# AQ-32. Save English Batch 03 results

ANNOTATION_FILE_03 = PROCESSED_DIR / "annotation_en_batch_03.csv"
QA_FILE_03 = PROCESSED_DIR / "qa_en_batch_03.csv"

final_en_batch_03.to_csv(
    ANNOTATION_FILE_03,
    index=False,
    encoding="utf-8-sig"
)

qa_results_en_03.to_csv(
    QA_FILE_03,
    index=False,
    encoding="utf-8-sig"
)

print("Saved annotation file:", ANNOTATION_FILE_03)
print("Saved QA file:", QA_FILE_03)

Saved annotation file: c:\Users\user\OneDrive\Documents\data-labeling-portfolio\01_multilingual_sentiment\data\processed\annotation_en_batch_03.csv
Saved QA file: c:\Users\user\OneDrive\Documents\data-labeling-portfolio\01_multilingual_sentiment\data\processed\qa_en_batch_03.csv


In [30]:
# AQ-33. Verify saved Batch 03 files

saved_annotation_03 = pd.read_csv(ANNOTATION_FILE_03)
saved_qa_03 = pd.read_csv(QA_FILE_03)

print("Annotation shape:", saved_annotation_03.shape)
print("QA shape:", saved_qa_03.shape)

print("\nMissing final labels:")
print(
    saved_annotation_03[
        ["final_polarity", "final_intensity"]
    ].isna().sum()
)

print("\nFinal polarity distribution:")
print(saved_annotation_03["final_polarity"].value_counts())

print("\nFinal intensity distribution:")
print(saved_annotation_03["final_intensity"].value_counts())

display(saved_annotation_03.head())

Annotation shape: (20, 6)
QA shape: (20, 9)

Missing final labels:
final_polarity     0
final_intensity    0
dtype: int64

Final polarity distribution:
final_polarity
Negative    20
Name: count, dtype: int64

Final intensity distribution:
final_intensity
Medium    13
High       6
Low        1
Name: count, dtype: int64


,review_id,review_title,review_body,final_polarity,final_intensity,language
0,en_0941733,Sizing,Ordered the size I always buy and it’s too sma...,Negative,Medium,en
1,en_0021765,Does not work!,This shampoo did not remove orange from my hai...,Negative,High,en
2,en_0422019,Too Big,Absolutely Giant. Order in the next size down....,Negative,High,en
3,en_0886190,Not safe due to chemical smells.,"Extremely strong chemical smell, made my pet r...",Negative,High,en
4,en_0375426,Didn’t do anything,I ordered this bc my 9 year old son had a wart...,Negative,Medium,en


## AQ-34. English Annotation Batch 04

English reviews #90–109 are annotated using **Annotation Guideline v0.4**.

This batch evaluates whether the newly added polarity edge-case rules
improve annotation consistency.

### Annotation Focus

- Assign only **Positive** or **Negative** polarity
- Use the review's **dominant evaluative direction**
- Do not create Mixed, Neutral, or Unclear labels when polarity can be determined
- Continue applying v0.3 intensity criteria

In [31]:
# AQ-34. Create English annotation batch #90-109

en_batch_04 = df_en.loc[
    90:109,
    ["review_id", "review_title", "review_body"]
].copy()

en_batch_04["polarity"] = ""
en_batch_04["intensity"] = ""

print("Batch size:", len(en_batch_04))

en_batch_04

Batch size: 20


,review_id,review_title,review_body,polarity,intensity
90,en_0460401,"Envelope doesn't seal, but forms are fine",The Envelope adhesive was applied to the wrong...,,
91,en_0608807,Fake REVIEWS,use fakespot.com to analyze the quality of the...,,
92,en_0724758,Some good...some the same contained in other c...,"L' Amour stories all rate an A+, except for on...",,
93,en_0227686,Is different than I ordered,Completely different cover than I ordered. I d...,,
94,en_0446730,Cute but little too big,Bear was alittle too big. The description of p...,,
95,en_0615055,Nice bowls with poor attachment features,After prying (and I mean prying to point of be...,,
96,en_0543284,Two Stars,Magnetic connections fail soon after using it....,,
97,en_0341678,Deformed but sturdy,The colors of the fiesta tortilla warmers were...,,
98,en_0056650,The color is way off,The color is way off and the picture doesn't f...,,
99,en_0170892,End of wand is weak,Cats loved it until the tip broke off it could...,,


In [32]:
# AQ-35. Display full review text for manual annotation

for idx, row in en_batch_04.iterrows():
    print(f"[{idx}]")
    print(f"Title: {row['review_title']}")
    print(f"Body : {row['review_body']}")
    print("-" * 80)

[90]
Title: Envelope doesn't seal, but forms are fine
Body : The Envelope adhesive was applied to the wrong areas so when you fold the envelope the 2 self sealing adhesives don't line up. I had to tape each envelope because even liquid envelope adhesive sealer didn't work because one side of it was hitting the self sealing adhesive. The pic shows and outline where the adhesive hits and how the 2 sides don't line up. The forms were good and worked with QuickBooks fine.
--------------------------------------------------------------------------------
[91]
Title: Fake REVIEWS
Body : use fakespot.com to analyze the quality of the reviews after I purchased the product. It is full of unreliable feedback
--------------------------------------------------------------------------------
[92]
Title: Some good...some the same contained in other collections.
Body : L' Amour stories all rate an A+, except for one of his earlier shorts. But the collections bother me. I have read several of the collect

## AQ-36. QA Review — English Batch 04

Batch 04 was annotated using **Annotation Guideline v0.4**.

This QA review evaluates whether the revised polarity rules improved
consistency while continuing to monitor sentiment intensity agreement.

In [33]:
# AQ-36. Record QA results — English Batch 04

qa_results_en_04 = pd.DataFrame({
    "review_id": en_batch_04["review_id"].values,

    "annotator_polarity": ["Negative"] * 20,

    "annotator_intensity": [
        "Medium", "High", "Medium", "Medium", "Medium",
        "Medium", "Medium", "Medium", "Medium", "Low",
        "Medium", "Medium", "Medium", "Medium", "Medium",
        "Medium", "Medium", "Medium", "Medium", "Low"
    ],

    "qa_polarity": ["Negative"] * 20,

    "qa_intensity": [
        "Medium", "High", "Medium", "Medium", "Medium",
        "Medium", "Medium", "Medium", "Medium", "Low",
        "Medium", "High", "Medium", "Medium", "High",
        "Medium", "High", "High", "Medium", "Low"
    ]
})

qa_results_en_04["polarity_agreement"] = (
    qa_results_en_04["annotator_polarity"]
    == qa_results_en_04["qa_polarity"]
)

qa_results_en_04["intensity_agreement"] = (
    qa_results_en_04["annotator_intensity"]
    == qa_results_en_04["qa_intensity"]
)

qa_results_en_04

,review_id,annotator_polarity,annotator_intensity,qa_polarity,qa_intensity,polarity_agreement,intensity_agreement
0,en_0460401,Negative,Medium,Negative,Medium,True,True
1,en_0608807,Negative,High,Negative,High,True,True
2,en_0724758,Negative,Medium,Negative,Medium,True,True
3,en_0227686,Negative,Medium,Negative,Medium,True,True
4,en_0446730,Negative,Medium,Negative,Medium,True,True
5,en_0615055,Negative,Medium,Negative,Medium,True,True
6,en_0543284,Negative,Medium,Negative,Medium,True,True
7,en_0341678,Negative,Medium,Negative,Medium,True,True
8,en_0056650,Negative,Medium,Negative,Medium,True,True
9,en_0170892,Negative,Low,Negative,Low,True,True


In [34]:
# AQ-37. Calculate Batch 04 QA metrics

polarity_agreement_04 = (
    qa_results_en_04["polarity_agreement"].mean() * 100
)

intensity_agreement_04 = (
    qa_results_en_04["intensity_agreement"].mean() * 100
)

exact_agreement_04 = (
    (
        qa_results_en_04["polarity_agreement"]
        & qa_results_en_04["intensity_agreement"]
    ).mean() * 100
)

print(f"Polarity agreement : {polarity_agreement_04:.1f}%")
print(f"Intensity agreement: {intensity_agreement_04:.1f}%")
print(f"Exact label agreement: {exact_agreement_04:.1f}%")

Polarity agreement : 100.0%
Intensity agreement: 80.0%
Exact label agreement: 80.0%


In [35]:
# AQ-38. Extract Batch 04 disagreements

intensity_disagreements_04 = qa_results_en_04[
    ~qa_results_en_04["intensity_agreement"]
].copy()

print(
    "Intensity disagreements:",
    len(intensity_disagreements_04)
)

print("\nDisagreement review IDs:")
print(intensity_disagreements_04["review_id"].tolist())

intensity_disagreements_04

Intensity disagreements: 4

Disagreement review IDs:
['en_0714896', 'en_0776613', 'en_0032500', 'en_0065166']


,review_id,annotator_polarity,annotator_intensity,qa_polarity,qa_intensity,polarity_agreement,intensity_agreement
11,en_0714896,Negative,Medium,Negative,High,True,False
14,en_0776613,Negative,Medium,Negative,High,True,False
16,en_0032500,Negative,Medium,Negative,High,True,False
17,en_0065166,Negative,Medium,Negative,High,True,False


## AQ-39. Final Label Adjudication — English Batch 04

Following QA review, four intensity disagreements were adjudicated.

Polarity achieved full agreement, while the remaining disagreements
involved the boundary between Medium and High intensity.

QA-reviewed labels are used as the final labels while preserving the
original annotator decisions in the QA audit dataset.

In [36]:
# AQ-39. Assign final labels — Batch 04

qa_results_en_04["final_polarity"] = qa_results_en_04["qa_polarity"]
qa_results_en_04["final_intensity"] = qa_results_en_04["qa_intensity"]

qa_results_en_04[
    [
        "review_id",
        "annotator_polarity",
        "annotator_intensity",
        "qa_polarity",
        "qa_intensity",
        "final_polarity",
        "final_intensity"
    ]
]

,review_id,annotator_polarity,annotator_intensity,qa_polarity,qa_intensity,final_polarity,final_intensity
0,en_0460401,Negative,Medium,Negative,Medium,Negative,Medium
1,en_0608807,Negative,High,Negative,High,Negative,High
2,en_0724758,Negative,Medium,Negative,Medium,Negative,Medium
3,en_0227686,Negative,Medium,Negative,Medium,Negative,Medium
4,en_0446730,Negative,Medium,Negative,Medium,Negative,Medium
5,en_0615055,Negative,Medium,Negative,Medium,Negative,Medium
6,en_0543284,Negative,Medium,Negative,Medium,Negative,Medium
7,en_0341678,Negative,Medium,Negative,Medium,Negative,Medium
8,en_0056650,Negative,Medium,Negative,Medium,Negative,Medium
9,en_0170892,Negative,Low,Negative,Low,Negative,Low


In [37]:
# AQ-40. Create finalized annotation dataset — Batch 04

final_en_batch_04 = en_batch_04[
    ["review_id", "review_title", "review_body"]
].merge(
    qa_results_en_04[
        ["review_id", "final_polarity", "final_intensity"]
    ],
    on="review_id",
    how="left"
)

final_en_batch_04["language"] = "en"

final_en_batch_04

,review_id,review_title,review_body,final_polarity,final_intensity,language
0,en_0460401,"Envelope doesn't seal, but forms are fine",The Envelope adhesive was applied to the wrong...,Negative,Medium,en
1,en_0608807,Fake REVIEWS,use fakespot.com to analyze the quality of the...,Negative,High,en
2,en_0724758,Some good...some the same contained in other c...,"L' Amour stories all rate an A+, except for on...",Negative,Medium,en
3,en_0227686,Is different than I ordered,Completely different cover than I ordered. I d...,Negative,Medium,en
4,en_0446730,Cute but little too big,Bear was alittle too big. The description of p...,Negative,Medium,en
5,en_0615055,Nice bowls with poor attachment features,After prying (and I mean prying to point of be...,Negative,Medium,en
6,en_0543284,Two Stars,Magnetic connections fail soon after using it....,Negative,Medium,en
7,en_0341678,Deformed but sturdy,The colors of the fiesta tortilla warmers were...,Negative,Medium,en
8,en_0056650,The color is way off,The color is way off and the picture doesn't f...,Negative,Medium,en
9,en_0170892,End of wand is weak,Cats loved it until the tip broke off it could...,Negative,Low,en


In [38]:
# AQ-41. Save English Batch 04 results

ANNOTATION_FILE_04 = PROCESSED_DIR / "annotation_en_batch_04.csv"
QA_FILE_04 = PROCESSED_DIR / "qa_en_batch_04.csv"

final_en_batch_04.to_csv(
    ANNOTATION_FILE_04,
    index=False,
    encoding="utf-8-sig"
)

qa_results_en_04.to_csv(
    QA_FILE_04,
    index=False,
    encoding="utf-8-sig"
)

print("Saved annotation file:", ANNOTATION_FILE_04)
print("Saved QA file:", QA_FILE_04)

Saved annotation file: c:\Users\user\OneDrive\Documents\data-labeling-portfolio\01_multilingual_sentiment\data\processed\annotation_en_batch_04.csv
Saved QA file: c:\Users\user\OneDrive\Documents\data-labeling-portfolio\01_multilingual_sentiment\data\processed\qa_en_batch_04.csv


In [39]:
# AQ-42. Verify saved Batch 04 files

saved_annotation_04 = pd.read_csv(ANNOTATION_FILE_04)
saved_qa_04 = pd.read_csv(QA_FILE_04)

print("Annotation shape:", saved_annotation_04.shape)
print("QA shape:", saved_qa_04.shape)

print("\nMissing final labels:")
print(
    saved_annotation_04[
        ["final_polarity", "final_intensity"]
    ].isna().sum()
)

print("\nFinal polarity distribution:")
print(saved_annotation_04["final_polarity"].value_counts())

print("\nFinal intensity distribution:")
print(saved_annotation_04["final_intensity"].value_counts())

display(saved_annotation_04.head())

Annotation shape: (20, 6)
QA shape: (20, 9)

Missing final labels:
final_polarity     0
final_intensity    0
dtype: int64

Final polarity distribution:
final_polarity
Negative    20
Name: count, dtype: int64

Final intensity distribution:
final_intensity
Medium    13
High       5
Low        2
Name: count, dtype: int64


,review_id,review_title,review_body,final_polarity,final_intensity,language
0,en_0460401,"Envelope doesn't seal, but forms are fine",The Envelope adhesive was applied to the wrong...,Negative,Medium,en
1,en_0608807,Fake REVIEWS,use fakespot.com to analyze the quality of the...,Negative,High,en
2,en_0724758,Some good...some the same contained in other c...,"L' Amour stories all rate an A+, except for on...",Negative,Medium,en
3,en_0227686,Is different than I ordered,Completely different cover than I ordered. I d...,Negative,Medium,en
4,en_0446730,Cute but little too big,Bear was alittle too big. The description of p...,Negative,Medium,en


## AQ-43. English Annotation QA Summary

This section summarizes QA performance across the first four English
production annotation batches.

A total of **80 reviews (#30–109)** were independently annotated,
reviewed, adjudicated, and finalized.

QA findings were used iteratively to refine the annotation guideline
and improve annotation consistency.

In [40]:
# AQ-43. Create QA summary across English batches

qa_summary_en = pd.DataFrame({
    "Batch": [
        "EN Batch 01",
        "EN Batch 02",
        "EN Batch 03",
        "EN Batch 04"
    ],
    "Reviews": [20, 20, 20, 20],
    "Guideline": [
        "v0.2",
        "v0.3",
        "v0.3",
        "v0.4"
    ],
    "Polarity Agreement (%)": [
        qa_results_en_01["polarity_agreement"].mean() * 100,
        qa_results_en_02["polarity_agreement"].mean() * 100,
        qa_results_en_03["polarity_agreement"].mean() * 100,
        qa_results_en_04["polarity_agreement"].mean() * 100
    ],
    "Intensity Agreement (%)": [
        qa_results_en_01["intensity_agreement"].mean() * 100,
        qa_results_en_02["intensity_agreement"].mean() * 100,
        qa_results_en_03["intensity_agreement"].mean() * 100,
        qa_results_en_04["intensity_agreement"].mean() * 100
    ],
    "Exact Agreement (%)": [
        (
            qa_results_en_01["polarity_agreement"]
            & qa_results_en_01["intensity_agreement"]
        ).mean() * 100,
        (
            qa_results_en_02["polarity_agreement"]
            & qa_results_en_02["intensity_agreement"]
        ).mean() * 100,
        (
            qa_results_en_03["polarity_agreement"]
            & qa_results_en_03["intensity_agreement"]
        ).mean() * 100,
        (
            qa_results_en_04["polarity_agreement"]
            & qa_results_en_04["intensity_agreement"]
        ).mean() * 100
    ]
})

qa_summary_en.round(1)

,Batch,Reviews,Guideline,Polarity Agreement (%),Intensity Agreement (%),Exact Agreement (%)
0,EN Batch 01,20,v0.2,100.0,45.0,45.0
1,EN Batch 02,20,v0.3,100.0,90.0,90.0
2,EN Batch 03,20,v0.3,70.0,80.0,55.0
3,EN Batch 04,20,v0.4,100.0,80.0,80.0


In [41]:
# AQ-44. Calculate overall English QA metrics

all_qa_en = pd.concat(
    [
        qa_results_en_01,
        qa_results_en_02,
        qa_results_en_03,
        qa_results_en_04
    ],
    ignore_index=True
)

overall_polarity = (
    all_qa_en["polarity_agreement"].mean() * 100
)

overall_intensity = (
    all_qa_en["intensity_agreement"].mean() * 100
)

overall_exact = (
    (
        all_qa_en["polarity_agreement"]
        & all_qa_en["intensity_agreement"]
    ).mean() * 100
)

print("Total reviewed:", len(all_qa_en))
print(f"Overall polarity agreement : {overall_polarity:.1f}%")
print(f"Overall intensity agreement: {overall_intensity:.1f}%")
print(f"Overall exact agreement    : {overall_exact:.1f}%")

Total reviewed: 80
Overall polarity agreement : 92.5%
Overall intensity agreement: 73.8%
Overall exact agreement    : 67.5%


## AQ-45. Key QA Findings

### 1. Intensity Calibration

English Batch 01 showed high polarity consistency but low intensity
agreement (**45%**).

QA analysis identified a systematic tendency to equate the severity of
a reported problem with the strength of sentiment expressed in the text.

Guideline v0.3 therefore clarified:

> **Problem severity ≠ Sentiment intensity**

After refinement, intensity agreement increased to **90% in Batch 02**
and remained at **80% in Batches 03 and 04**.

### 2. Polarity Edge Cases

Batch 03 revealed a second annotation issue involving reviews containing
both positive and negative statements.

The annotator introduced Mixed, Neutral, and Unclear labels, resulting
in polarity agreement of **70%**.

Guideline v0.4 therefore introduced the principle:

> **Mixed content ≠ Mixed label**

Under the binary schema, polarity is determined using the review's
dominant evaluative direction.

Batch 04 subsequently achieved **100% polarity agreement**.

### 3. Overall QA

Across the first **80 English production reviews**:

- Polarity agreement: **92.5%**
- Intensity agreement: **73.8%**
- Exact label agreement: **67.5%**

These results demonstrate an iterative QA workflow in which annotation
disagreements are analyzed, converted into guideline improvements, and
evaluated in subsequent annotation batches.

Because each batch contains only 20 reviews, batch-level improvements
are treated as evidence of improved consistency rather than statistically
conclusive performance changes.